docs/architecture/adr-039-add-support-for-wasm-light-clients.md

```diff
@@ -0,0 +1,107 @@
+hard-fork event.
+
+## Context
+Currently in the SDK, light clients are defined as part of the codebase and are implemented as submodules under
+`/x/ibc/light-clients/`.
+
+Adding support for new light client or update an existing light client in the event of security
+issue or consensus update is multi-step process:
+
+1. Light client integration with SDK: IBC light clients are defined as part of the codebase and are implemented as
+submodules under `/x/ibc/light-clients/`. To add support for new light client or update an existing light client in the
+event of security issue or consensus update, we need to modify the codebase and integrate it in numerous places.
+
+2. Governance voting: Adding new light client implementations require governance support and is expensive: This is
+not ideal as chain governance is gatekeeper for new light client implementations getting added. If a small community
+want support for light client X, they may not be able to convince governance to support it.
+
+3. Validator upgrade: After governance voting succeeds, validators need to upgrade their nodes in order to enable new
+IBC light client implementation. This is both time consuming and error prone.
+   
+Another problem stemming from above process is, if a chain want to upgrade its own consensus, it will need to convince every chain
+or hub connected to it to upgrade its light client in order to stay connected. Due to time consuming process required
+to upgrade light client, a chain with lots of connection need to be disconnected for quite some time after upgrading 
+its consensus, which can be very expensive.
+
+We are proposing simplifying this workflow by integrating a WASM light client module which makes adding support for
+a new light client a simple transaction. The light client bytecode, written in Wasm-compilable Rust, runs inside a Wasmer
+VM. The Wasm light client submodule exposes a proxy light client interface that routes incoming messages to the
+appropriate handler function, inside the Wasm VM for execution.
+
+With WASM light client module, anybody can add new IBC light client in the form of WASM bytecode (provided they are able to pay huge gas fee for the transaction)
+as well as instantiate clients using any created client type. This allows any chain to update its own light client in other chains
+without going through steps outlined above.
+
+
+## Decision
+
+We decided to use WASM light client module as a generic light client which will interface with the actual light client
+uploaded as WASM bytecode. This will require changing client selection method to allow any client if the client type
+has prefix of `wasm/`.
+
+```go
+// IsAllowedClient checks if the given client type is registered on the allowlist.
+func (p Params) IsAllowedClient(clientType string) bool {
+	for _, allowedClient := range p.AllowedClients {
+		if allowedClient == clientType {
+			return true
+		}
+	}
+	return false || (p.AreWASMClientsAllowed && isWASMClient(clientType))
```

Comment:
I think maybe we should have some default validation function here. This hardcodes WASM clients into all chains which want to use IBC (even if they don't want to use WASM or support it). 

A default validator function could do the above logic, while allowing the validator function to be replaced with different custom logic (for example, in another complex light client implementation)

I think it'd look something like:
```go
func ValidateClientType(p Params, clientType string) bool {
        for _, allowedClient := range p.AllowedClients {
		if allowedClient == clientType {
			return true
		}
	}
	return false || (p.AreWASMClientsAllowed && isWASMClient(clientType))
}

// IsAllowedClient calls the ValidateClientType function to determine if the client type is allowed.
func (p Params) IsAllowedClient(clientType string) bool {
    return ValidateClientType(p, clientType)
}
```
It's probably possible to construct this in a cleaner way